### PR TITLE
build: setting up Rollup for bundling

### PR DESCRIPTION
**Pull request by zchenwei**
#### Description of changes

For tree shaking improvement, using Rollup as new bundler for `@aws-amplify/ui` and `@aws-amplify/ui-react`. https://quip-amazon.com/ZxjgAFCqQbXQ/Tree-shaking-improvement

#### Issue #, if available

`@aws-amplify/ui` and `@aws-amplify/ui-react` are non-treeshakable - https://quip-amazon.com/WbtjAEJ1ZsID/Tree-shaking-woes

#### Description of how you validated changes

Testing against some popular build tools

- [x] webpack

```jsx
// index.tsx
import * as React from 'react';
import ReactDOM from 'react-dom';
import '@aws-amplify/ui-react/styles.css';

import { App } from './App';

ReactDOM.render(<App />, document.getElementById('root'));

// App.tsx
import * as React from 'react';
import { Loader } from '@aws-amplify/ui-react';

export const App = () => {
  return <Loader />;
};
```
main.js size
| Before     | After |
| ----------- | ----------- |
|   1.7M   | 161.1k      |

- [x] create-react-app

```jsx
// index.js
import ReactDOM from 'react-dom';
import App from './App';
import Amplify from 'aws-amplify';
import awsExports from './aws-exports';
Amplify.configure(awsExports);

ReactDOM.render(<App />, document.getElementById('root'));

// App.js

import { Authenticator } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';

function App() {
  return (
    <Authenticator>
      {({ signOut, user }) => (
        <main>
          <h1>Hello {user.username}</h1>
          <button onClick={signOut}>Sign out</button>
        </main>
      )}
    </Authenticator>
  );
}

export default App;
```

main.js size(gzip)
| Before     | After |
| ----------- | ----------- |
|   499.85kB   | 261.46kB  |

- [x] Vite

```jsx
// main.jsx
import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';

ReactDOM.render(<App />, document.getElementById('root'));

// App.jsx
import { Loader } from '@aws-amplify/ui-react';
import '@aws-amplify/ui-react/styles.css';

function App() {
  return <Loader />;
}

export default App;
```

vendor.js size(gzip)
| Before     | After |
| ----------- | ----------- |
|  446.69 KiB  | 58.29 KiB  |

#### Checklist

- [x] PR description included
- [x] `yarn test` passes

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/ui/rollup.config.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  },
+  // ESM config
+  {
+    input: ['src/index.ts'],
+    output: {
+      dir: 'dist/esm',
+      format: 'es',
+      preserveModules: true,
+      preserveModulesRoot: 'src',
+      sourcemap: true,
```
Comment (ErikCH) [inline packages/ui/rollup.config.js:26]: So will sourcemaps be bundled in for production?

File: packages/react/rollup.config.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+const config = defineConfig([
+  // CJS config
+  {
+    input: ['src/index.tsx', 'src/internal.tsx', 'src/legacy.tsx'],
+    output: {
+      dir: 'dist',
+      format: 'cjs',
+      sourcemap: true,
+    },
+    plugins: [typescript({ declarationDir: 'dist/types' }), terser()],
```
Comment (ErikCH) [inline packages/react/rollup.config.js:16]: What does terser do?

File: packages/ui/rollup.config.js (hunk 1 above)
Comment (zchenwei) [reply packages/ui/rollup.config.js:26]: Yes, sourcemaps will be bundled for prod per our discussion and see if anyone complains about that.

File: packages/react/rollup.config.js (hunk 2 above)
Comment (zchenwei) [reply packages/react/rollup.config.js:16]: terser is a code compressor. Using it to minify library size

File: packages/ui/rollup.config.js (hunk 1 above)
Comment (calebpollman) [reply packages/ui/rollup.config.js:26]: How much impact does that have on bundle size?

**Review by wlee221 — COMMENTED** [review]
Small Nit: https://github.com/aws-amplify/amplify-ui/blob/c070cc8a155a351beaf2d5932d85d965ad8e5d26/package.json#L80

should be removed 😛

File: packages/react/rollup.config.js (hunk 2 above)
Comment (ErikCH) [reply packages/react/rollup.config.js:16]: 👍

**Review by ErikCH — APPROVED** [review]
👍

File: packages/ui/rollup.config.js (hunk 1 above)
Comment (zchenwei) [reply packages/ui/rollup.config.js:26]: > Small Nit:
> 
> https://github.com/aws-amplify/amplify-ui/blob/c070cc8a155a351beaf2d5932d85d965ad8e5d26/package.json#L80
> 
> should be removed 😛

That's a good catch

**Comment by zchenwei** [comment]
> Small Nit:
> 
> https://github.com/aws-amplify/amplify-ui/blob/c070cc8a155a351beaf2d5932d85d965ad8e5d26/package.json#L80
> 
> should be removed 😛

That's a good catch

File: packages/ui/rollup.config.js (hunk 1 above)
Comment (zchenwei) [reply packages/ui/rollup.config.js:26]: Almost 30% increase I think.

**Comment by reesscot** [comment]
Should we add a changeset? At a minimum I would think this would be a patch, but I could also see it being a minor bump given its impact on bundle sizes!

**Comment by wlee221** [comment]
> Should we add a changeset?

You're totally right, we should! I'll let @zchenwei decide patch vs minor 😈

**Review by reesscot — APPROVED** [review]
LGTM! 

I did the following smoke testing

1. Pulled your branch
2. npm publish to local verdaccio npm registry
3. npm install into a CRA app
4. Smoke tested app and didn't see any console errors.

File: packages/react/rollup.config.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      format: 'es',
+      preserveModules: true,
+      preserveModulesRoot: 'src',
+      sourcemap: true,
+    },
+    plugins: [typescript({ outDir: 'dist/esm', declaration: false }), terser()],
+  },
+  // CSS config
+  {
+    input: 'src/styles.ts',
```
[the remark this answers is not in the document]
Comment (zchenwei) [reply packages/react/rollup.config.js:32]: Yup, that way we can separate the bundling for styling and source code.

[the remark this answers is not in the document]
Comment (zchenwei) [reply packages/react/rollup.config.js:32]: Otherwise it will emit CSS twice. I think we forgot to clean it up somehow

**Comment by zchenwei** [comment]
> > Should we add a changeset?
> 
> You're totally right, we should! I'll let @zchenwei decide patch vs minor 😈

Let's do a minor bump 😈

**Review by wlee221 — APPROVED** [review]

**Comment by ericclemmons** [comment]
Can the Quip docs be made public? This means that esbuild produces unoptimizable bundles and could affect other tools like Vite for libraries.

**Comment by zchenwei** [comment]
> Can the Quip docs be made public? This means that esbuild produces unoptimizable bundles and could affect other tools like Vite for libraries.

That's a good call. I have no answer if I can make it public but that reminds me I should remove the internal link 🤣

**Pull request merged**